Backport of ranking evaluation API (#27478)

cbuescher opened the pull request:
This is the backport of the ranking evaluation feature (#27478) to the 6.x branches.
In addition to the code commited on master I added back support for types in the request URL so
users can use the API to evaluate on indices created with multiple types also. However, only one
type per request is supported to avoid having to re-introduce a type parameter in the rated documents as well, otherwise an error is thrown.

Commit:
Add ranking evaluation API

This commit adds a new module that provides an endpoint that can be used to
evaluate search ranking results.

Relates to #19195

Commit:
Add skip versions for rank_eval yaml tests

jimczi commented on docs/reference/search/rank-eval.asciidoc:
```diff
@@ -7,9 +7,9 @@ list or manually rated documents, the `_rank_eval` endpoint calculates and
 
-experimental[The ranking evaluation API is new and may change in non-backwards compatible ways in the future,
-even on minor versions updates.]
+experimental[The ranking evaluation API is new and may change in non-backwards compatible ways in the future, even on minor versions updates.]
```
I missed that in the first review. I don't think you need to add anything to `experimental[]`, the message is auto generated with the appropriate content.

cbuescher replied:
There already was an experimental tag on master but it wasn't working as expected since I had a line break in the text. I added the custom text since the auto generated message reads `"This functionality is experimental and may be changed or removed completely in a future release. Elastic will take a best effort approach to fix any issues, but experimental features are not subject to the support SLA of official GA features."` which is fine, but I wanted to include the notion about changes in minor version updates since its likely we will slighty change the DSL a bit in the beginning.

jimczi reviewed (changes requested):
I don't think it is necessary to add the support for a single type like this. What is the benefit ? You only accept a single type per request so it's the same than without ?

cbuescher commented:
> I don't think it is necessary to add the support for a single type like this. What is the benefit ?

The idea is to allow users that have a pre 6.x index with multiple types to select one type to use in ranking evaluation. It is necessary because when specifying the ratings in the request, a document it identified only by using its `index` and `id` (otherwise I would have to re-introduce the `type` as an identifier there which makes the requests unnecessarily bigger). If the user wouldn't be able to pick a type in the Url, some documents in different types might have the same index/id but still might be returned by a search across all types, making the attached rating ambiguous.

jimczi commented:
> If the user wouldn't be able to pick a type in the Url, some documents in different types might have the same index/id but still might be returned by a search across all types, making the attached rating ambiguous.

They can add a filter in the query that restricts the search to the `_type` they are interested in ?
I don't think we should add an endpoint that uses `_type` to a new API like this. It can be a small note in the docs that explains how to filter types using the query DSL.

cbuescher commented:
> They can add a filter in the query that restricts the search to the _type they are interested in ?

Sold! That didn't cross my mind when I was talking to @clintongormley about it, but thats much better. Thanks for suggesting this, I will add a short note to the docs instead and remove the type support again from this branch.

cbuescher commented:
@jimczi I reverted the commit that added the type support and added a note to the docs to show how working with old indices with multiple types can work.

jimczi commented on docs/reference/search/rank-eval.asciidoc:
```diff
@@ -86,6 +87,11 @@ the rating of the documents relevance with regards to this search request
 
 A document `rating` can be any integer value that expresses the relevance of the document on a user defined scale. For some of the metrics, just giving a binary rating (e.g. `0` for irrelevant and `1` for relevant) will be sufficient, other metrics can use a more fine grained scale.  
 
+NOTE: To use the ranking evaluation API with indices that use multiple types, you should add a filter on the `_type` field to
```
I think it should appear on the same line to not break the NOTE.

cbuescher replied:
I just checked, multi line seems to work.

jimczi commented on docs/reference/search/rank-eval.asciidoc:
```diff
@@ -86,6 +87,11 @@ the rating of the documents relevance with regards to this search request
 A document `rating` can be any integer value that expresses the relevance of the document on a user defined scale. For some of the metrics, just giving a binary rating (e.g. `0` for irrelevant and `1` for relevant) will be sufficient, other metrics can use a more fine grained scale.  
 
+NOTE: To use the ranking evaluation API with indices that use multiple types, you should add a filter on the `_type` field to
+the query in the request. Otherwise, if your index uses multiple types with the same id, the privided 
```
nit: s/privided/provided/

cbuescher replied:
changed that.

jimczi reviewed (approved):
LGTM

Commit:
Corrected typo

cbuescher commented:
@elasticmachine test this again please
